### PR TITLE
feat: custom navbarItem types (workaround)

### PR DESCRIPTION
**Pull request by slorber**
## Motivation

This implements a temporary workaround to support custom navbar item types:

- Provide items with `type: "custom-x"` in navbar items (+ arbitrary props)
- Swizzle `@theme/NavbarItem/ComponentTypes` to register your `custom-x` component

Note this is a temporary workaround until we provide a better API to register custom navbar item types.

See also: https://github.com/facebook/docusaurus/issues/7227#issuecomment-1106336591

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

dogfood + preview + validation tests

The test section of the website will show a custom navbar item:

https://deploy-preview-7231--docusaurus-2.netlify.app/tests/docs/

CleanShot 2022-04-22 at 19 57 57@2x

### CODE REVIEW

**Comment by slorber** [comment]
cc @jaredcbaum , looks relevant for what you did on Ionic doc here https://github.com/ionic-team/ionic-docs/blob/main/src/theme/NavbarItem/index.tsx

**Comment by jaredcbaum** [comment]
I attempted another workaround in the latest batch of updates to the ionic docs. This abstracts the validation logic to the Docusaurus config file. I was/am attempting to abstract a common theme that could also have custom schemas based on which site it was used on. I only got far enough for the navbar but theoretically it could be used for other schemas.

https://github.com/ionic-team/ionic-docs/blob/feat/plugin-abstraction/docusaurus.config.js#L211-L239
https://github.com/ionic-team/docusaurus-theme-classic-wrapper/blob/main/src/index.js#L28-L32

I think this is dancing around a bigger problem though, which is custom navbar parts. It seems as though this is in the works, but figured I'd post in case anyone was looking for an immediate workaround.

**Comment by jaredcbaum** [comment]
Also, just an idea but I think it would work well to have a generated list of the "native" NavbarItems from the original theme and anything else would be considered custom. In theory it could eliminate the need to prefix with "custom-"

**Comment by slorber** [comment]
Thanks @jaredcbaum, another workaround remains useful until we have proper support for this 👍 

> Also, just an idea but I think it would work well to have a generated list of the "native" NavbarItems from the original theme and anything else would be considered custom. In theory it could eliminate the need to prefix with "custom-"

We could as well allow unknown navbar item types in our navigation instead of constraining to `custom-*` pattern. But that also means we don't catch typos anymore, which we likely want to avoid for regular users that don't implement custom navbar items.

For now prefixing things with `custom-*` is likely good enough and would catch typos too.

File: packages/docusaurus-theme-classic/src/theme/NavbarItem/index.tsx [hunk 1]
```diff
@@ -65,5 +34,6 @@ export default function NavbarItem({type, ...props}: Props): JSX.Element {
     (props as DropdownNavbarItemProps).items !== undefined,
   );
   const NavbarItemComponent = getNavbarItemComponent(componentType);
+  // @ts-expect-error: how to type this?
   return <NavbarItemComponent {...props} />;
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme/NavbarItem/index.tsx:38]:
```suggestion
  return <NavbarItemComponent {...(props as never)} />;
```

Will break in the next TS version, but that's something to worry about later. Polymorphic components are hard to type.

**Comment by medmin** [comment]
may i ask what's the status of this new feature?

I just tried to swizzle `@theme/NavbarItem/ComponentTypes` , but I cannot find it from the swizzling list.

**Pull request merged**